### PR TITLE
Allow UpdateWorkflowExecution against a completed workflow

### PR DESCRIPTION
**Pull request by stephanos**
## What changed?

Return previous Update result even if WF is already closed.

## Why?

Closing https://github.com/temporalio/temporal/issues/6085

## How did you test it?

Added functional test.

### CODE REVIEW

File: service/history/api/updateworkflow/api.go [hunk 1]
```diff
@@ -132,12 +132,21 @@ func (u *Updater) ApplyRequest(
...
 	ms workflow.MutableState,
 ) (*api.UpdateWorkflowAction, error) {
+	updateID := u.req.GetRequest().GetRequest().GetMeta().GetUpdateId()
+
 	if !ms.IsWorkflowExecutionRunning() {
+		// If the WF is not running anymore, use an existing Update, if it exists for the requested ID.
+		// This ensures that repeated Update requests with the same ID see the same result.
+		if u.upd = updateReg.Find(ctx, updateID); u.upd != nil {
+			return &api.UpdateWorkflowAction{Noop: true}, nil
+		}
```
Comment (stephanos) [inline service/history/api/updateworkflow/api.go:142]: By setting `upd`, the regular success codepath can be re-used.

**Comment by dandavison** [comment]
If a client submits an update request with `waitStage=Completed`, and the update ID exists, but the workflow has already completed with the update at `stage=Accepted`, I think the request will time out, right? It feels to me like the client should get an immediate `Workflow execution already completed` error rather than a timeout (since the client may know perfectly well that that this is the situation and that the update can never complete, so it would look a bit odd for the long-poll to block).

**Comment by stephanos** [comment]
You're probably right. Let me add a functional test for this case; and then figure out how make the necessary change.

**Review by dandavison — APPROVED** [review]
LGTM!

File: service/history/api/updateworkflow/api.go [hunk 2]
```diff
@@ -257,7 +267,8 @@ func (u *Updater) OnSuccess(
 			// If subsequent attempt succeeds within current context timeout, caller of this API will get a valid response.
 			err = nil
 		}
-	} else {
+	default:
+		// Normal WFT was created.
```
Comment (alexshtin) [inline service/history/api/updateworkflow/api.go:271]: I believe there is no such case any more.

File: tests/update_workflow.go [hunk 3]
```diff
@@ -863,6 +864,162 @@ func (s *FunctionalSuite) TestUpdateWorkflow_RunningWorkflowTask_NewNotEmptySpec
  14 WorkflowExecutionCompleted`, events)
 }
 
+func (s *FunctionalSuite) TestUpdateWorkflow_CompletedWorkflow() {
```
Comment (alexshtin) [inline tests/update_workflow.go:867]: Test need to be adjust with #6258.

Comment (stephanos) [reply tests/update_workflow.go:867]: I can fix it once https://github.com/temporalio/temporal/pull/6258 is merged 👍

File: service/history/api/updateworkflow/api.go (hunk 2 above)
Comment (stephanos) [reply service/history/api/updateworkflow/api.go:271]: 👍 I filed a separate ticket to address this

File: tests/update_workflow.go [hunk 4]
```diff
@@ -157,8 +158,12 @@ func (s *FunctionalSuite) waitUpdateAdmitted(tv *testvars.TestVars, updateID str
 			s.GreaterOrEqual(pollResp.Stage, enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ADMITTED)
 			return true
 		}
-		var notFoundErr *serviceerror.NotFound
-		s.ErrorAs(pollErr, &notFoundErr) // poll beat send in race
+		if pollErr.Error() != fmt.Sprintf("update \"%v\" not found", tv.UpdateID(updateID)) {
+			s.T().Log("received error from Update poll: ", pollErr)
+			return true
+		}
```
Comment (stephanos) [inline tests/update_workflow.go:164]: Added this to allow `workflow execution already completed` (also a NotFound) error from stopping the polling.

File: service/history/api/updateworkflow/api.go [hunk 5]
```diff
@@ -257,8 +260,6 @@ func (u *Updater) OnSuccess(
 			// If subsequent attempt succeeds within current context timeout, caller of this API will get a valid response.
 			err = nil
 		}
-	} else {
-		metrics.WorkflowExecutionUpdateNormalWorkflowTask.With(u.shardCtx.GetMetricsHandler()).Record(1)
```
Comment (stephanos) [inline service/history/api/updateworkflow/api.go:261]: Removed this and will put it in the right spot in a follow-up PR.

File: service/history/api/updateworkflow/api.go [hunk 6]
```diff
@@ -132,12 +131,21 @@ func (u *Updater) ApplyRequest(
...
+			return &api.UpdateWorkflowAction{Noop: true}, nil
+		}
 		return nil, consts.ErrWorkflowCompleted
 	}
+
 	u.wfKey = ms.GetWorkflowKey()
 
-	if u.req.GetRequest().GetFirstExecutionRunId() != "" && ms.GetExecutionInfo().GetFirstExecutionRunId() != u.req.GetRequest().GetFirstExecutionRunId() {
+	if u.req.GetRequest().GetFirstExecutionRunId() != "" &&
+		ms.GetExecutionInfo().GetFirstExecutionRunId() != u.req.GetRequest().GetFirstExecutionRunId() {
```
Comment (alexshtin) [inline service/history/api/updateworkflow/api.go:148]: I believe this check needs to be done before `!ms.IsWorkflowExecutionRunning()` check now.

**Review by alexshtin — APPROVED** [review]

File: tests/update_workflow.go (hunk 4 above)
Comment (alexshtin) [reply tests/update_workflow.go:164]: Can you leave type checking too? First check type then message.

File: tests/update_workflow.go [hunk 7]
```diff
@@ -157,8 +158,12 @@ func (s *FunctionalSuite) waitUpdateAdmitted(tv *testvars.TestVars, updateID str
 			s.GreaterOrEqual(pollResp.Stage, enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ADMITTED)
 			return true
 		}
-		var notFoundErr *serviceerror.NotFound
-		s.ErrorAs(pollErr, &notFoundErr) // poll beat send in race
+		if pollErr.Error() != fmt.Sprintf("update \"%v\" not found", tv.UpdateID(updateID)) {
```
Comment (alexshtin) [inline tests/update_workflow.go:161]:
```suggestion
		if pollErr.Error() != fmt.Sprintf("update %q not found", tv.UpdateID(updateID)) {
```
?

Comment (stephanos) [reply tests/update_workflow.go:161]: 👍 Nice, didn't know this one!

File: service/history/api/updateworkflow/api.go (hunk 6 above)
Comment (stephanos) [reply service/history/api/updateworkflow/api.go]: 👍 agreed!

**Pull request merged**